runtime/runtime-params-estimator/src/gas_cost.rs

```diff
@@ -271,3 +316,148 @@ impl GasCost {
+    }
+}
+#[test]
+fn least_squares_method_gas_cost_time_neg_factor() {
+    let xs = [10, 20, 30];
+
+    let ys =
+        [GasCost::new_time_based(990), GasCost::new_time_based(980), GasCost::new_time_based(970)];
+
+    match GasCost::least_squares_method_gas_cost(&xs, &ys) {
+        Err(((a_pos, b_pos), (a_neg, b_neg))) => {
+            assert_eq!(a_pos, GasCost::new_time_based(1000));
+            assert_eq!(a_neg, GasCost::new_time_based(0));
+
+            assert_eq!(b_pos, GasCost::new_time_based(0));
+            assert_eq!(b_neg, GasCost::new_time_based(1));
+        }
+        Ok(_) => panic!("Least squares got no negative values where it should have."),
+    }
+}
+
+#[test]
+fn least_squares_method_gas_cost_icount_ok() {
+    let xs = [10, 20, 30];
+
+    let ys = [
+        GasCost::new_icount_based(10_050, 20_060, 30_070),
+        GasCost::new_icount_based(20_050, 40_060, 60_070),
+        GasCost::new_icount_based(30_050, 60_060, 90_070),
+    ];
+
+    match GasCost::least_squares_method_gas_cost(&xs, &ys) {
+        Ok((a, b)) => {
+            assert_eq!(a, GasCost::new_icount_based(50, 60, 70));
+            assert_eq!(b, GasCost::new_icount_based(1000, 2000, 3000));
+        }
+        Err(((_, _), (a, b))) => {
+            panic!("Least squares got negative values where it shouldn't. Negative parts: a={:?} b={:?}", a, b)
+        }
+    }
+}
+
+#[test]
+fn least_squares_method_gas_cost_icount_neg_base() {
+    let xs = [10, 20, 30];
+
+    let ys = [
+        GasCost::new_icount_based(9_950, 19_960, 29_970),
+        GasCost::new_icount_based(19_950, 39_960, 59_970),
+        GasCost::new_icount_based(29_950, 59_960, 89_970),
+    ];
+
+    match GasCost::least_squares_method_gas_cost(&xs, &ys) {
+        Err(((a_pos, b_pos), (a_neg, b_neg))) => {
+            assert_eq!(a_pos, GasCost::new_icount_based(0, 0, 0));
+            assert_eq!(a_neg, GasCost::new_icount_based(50, 40, 30));
+
+            assert_eq!(b_pos, GasCost::new_icount_based(1000, 2000, 3000));
+            assert_eq!(b_neg, GasCost::new_icount_based(0, 0, 0));
+        }
+        Ok(_) => panic!("Least squares got no negative values where it should have."),
+    }
+}
+#[test]
+fn least_squares_method_gas_cost_icount_neg_factor() {
+    let xs = [10, 20, 30];
+
+    let ys = [
+        GasCost::new_icount_based(990, 981, 972),
+        GasCost::new_icount_based(980, 961, 942),
+        GasCost::new_icount_based(970, 941, 912),
+    ];
+
+    match GasCost::least_squares_method_gas_cost(&xs, &ys) {
+        Err(((a_pos, b_pos), (a_neg, b_neg))) => {
+            assert_eq!(a_pos, GasCost::new_icount_based(1000, 1001, 1002));
+            assert_eq!(a_neg, GasCost::new_icount_based(0, 0, 0));
+
+            assert_eq!(b_pos, GasCost::new_icount_based(0, 0, 0));
+            assert_eq!(b_neg, GasCost::new_icount_based(1, 2, 3));
+        }
+        Ok(_) => panic!("Least squares got no negative values where it should have."),
+    }
+}
```

Comment:
I see a minor smell here -- all these tests are coded directly to the API of the type, so, if the API of the type changes, you need to change every test. For example, if we move from returning `Result`s to using `uncertain` flag, you'd have to change every match. I think it's usually best to make the tests more data driven, by extracting a single driver function which interacts with the API, like this

```rust
#[cfg(test)]
mod tests {
  #[track_caller]
  fn check(xs: &[u64], ys: &[GasCost], expected: Result<(GasCost, GasCost), ()>) {}
  #[test]
  fn least_squares_method_gas_cost_icount_ok() {
    check(&[10, 20, 30], [
        GasCost::new_icount_based(990, 981, 972),
        GasCost::new_icount_based(980, 961, 942),
        GasCost::new_icount_based(970, 941, 912),
    ], Ok((GasCost::new_icount_based(50, 60, 70), GasCost::new_icount_based(50, 60, 70))))
  }
}
```

Reply:
Oh, and yeah, I think it's better to default to having a `tests module -- that way you can have additional imports and heplers. Theres' `tmod` snippet in both IntelliJ and RA to generate that.

Reply:
Thanks for the suggestions, definitely makes sense to do as you write. I will work on the change

Reply:
This should be addressed with the newest commit, if I understood your suggestions right